crates/bevy_pbr/src/render/shadow_sampling.wgsl

```diff
@@ -0,0 +1,134 @@
+#define_import_path bevy_pbr::shadow_sampling
+
+#import bevy_pbr::mesh_view_bindings as view_bindings
+#import bevy_pbr::utils PI
+
+fn sample_shadow_map_hardware(light_local: vec2<f32>, depth: f32, array_index: i32) -> f32 {
+    // Do the lookup, using HW 2x2 PCF and comparison
+    // NOTE: Due to non-uniform control flow above, we must use the level variant of the texture
+    // sampler to avoid use of implicit derivatives causing possible undefined behavior.
```

Comment:
I think now this is split into its own function that this is no longer true. I have some memory of the control flow only applying within a function scope. Though I also remember thinking that was weird given it sounds like shader compilers generally inline everything? @cwfitzgerald - is the non-uniform control flow for the entire execution of the shader after entering the region of non-uniform control flow? Or is it scoped to functions? Or is the validation scoped to functions but the rule does apply?

Reply:
cwfitzgerald confirmed that naga should emit validation errors if the texture sampling is done inside non-uniform control flow even if the code calls to another function.

Still, I think the `Level` variant makes sense regardless for shadow maps as they are not mipmapped and don't need gradients / mip level calculation to be done.